src/javascript/dimensions-metrics-explorer/components/columnGroupList.tsx

```diff
@@ -0,0 +1,346 @@
+        columns={columns.filter(
+          column => column.attributes.type === "DIMENSION"
+        )}
+        allowableCubes={allowableCubes}
+        cubes={cubes}
+        selectColumn={selectColumn}
+        selectedColumns={selectedColumns}
+      />
+      <ColumnSubgroup
+        name="Metrics"
+        columns={columns.filter(column => column.attributes.type === "METRIC")}
+        allowableCubes={allowableCubes}
+        cubes={cubes}
+        selectColumn={selectColumn}
+        selectedColumns={selectedColumns}
+      />
+    </div>
+  </div>
+);
+
+const ColumnGroupList: React.FC<{
+  allowDeprecated: boolean;
+  searchTerms: string[];
+  onlySegments: boolean;
+  cubes: CubesByColumn;
+  columns: Column[];
+}> = ({ allowDeprecated, searchTerms, onlySegments, columns, cubes }) => {
+  // Get the grouped columns
+  const groupedColumns = React.useMemo(() => {
+    let filteredColumns: Column[] = columns;
+
+    if (!allowDeprecated) {
+      filteredColumns = filteredColumns.filter(
+        column => column.attributes.status != "DEPRECATED"
+      );
+    }
+
+    if (onlySegments) {
+      filteredColumns = filteredColumns.filter(
+        column => column.attributes.allowedInSegments === "true"
+      );
+    }
+
+    filteredColumns = filteredColumns.filter(column =>
+      searchTerms.every(
+        term =>
+          column.id.toLowerCase().indexOf(term) != -1 ||
+          column.attributes.uiName.toLowerCase().indexOf(term) != -1
+      )
+    );
+
+    // JS Sets guarantee insertion order is preserved, which is important
+    // because the key order in this groupBy determines the order that
+    // they appear in the UI.
+    return groupBy(filteredColumns, column => column.attributes.group);
+  }, [columns, searchTerms, allowDeprecated, onlySegments]);
+
+  // Set of column groups that are currently expanded
+  const [open, setOpen] = React.useState<Set<string>>(() => Set());
+
+  // Expand/Collapse callbacks
+  const toggleGroupOpen = React.useCallback(
+    (group: string) =>
+      setOpen(oldOpen =>
+        oldOpen.contains(group) ? oldOpen.remove(group) : oldOpen.add(group)
+      ),
+    [setOpen]
+  );
+
+  const collapseAll = React.useCallback(() => setOpen(Set()), [setOpen]);
+
+  const expandAll = React.useCallback(
+    () => setOpen(Set.fromKeys(groupedColumns)),
+    [setOpen, groupedColumns]
+  );
+
+  // When a search term is entered, auto-expand all groups. When the search
+  // term is empty, auto-collapse all groups.
+  // KNOWN BUG: this effect should only run when searchTerms changes,
+  // but currently it also runs when expandAll changes, which happens
+  // when a checkbox is checked. Not sure how to fix this. Maybe a ref?
+  /*
+  React.useEffect(() => {
+    if (searchTerms.length === 0) {
+      collapseAll();
+    } else {
+      expandAll();
+    }
+  }, [searchTerms.length === 0, collapseAll, expandAll]);
+  */
+
+  // selectedColumns is a dictionary mapping each column to its selected
```

Comment:
It's not really a dictionary, I'd update this comment to mention a Set was chosen because you need to do `.intersection`, also, it's probably better to do `new Set()` so it's easier to follow what the initial value is.

Reply:
I'll update the column, but immutable actually doesn't like `new Set`